### ISSUE TITLE
InvalidCastException in IL2CPP builds (Unity 2021.2)

### ISSUE THREAD

**Issue by samlletas** [DavidF-Dev/Unity-DeveloperConsole#1]
After updating from Unity 2021.1 to 2021.2 the following exception is thrown in builds (both development and release):

```
InvalidCastException: Object must implement IConvertible.
  at System.Convert.ChangeType (System.Object value, System.Type conversionType, System.IFormatProvider provider) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Runtime.Serialization.SerializationInfo.GetValue (System.String name, System.Type type) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].OnDeserialization (System.Object sender) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize (System.Runtime.Remoting.Messaging.HeaderHandler handler, System.Runtime.Serialization.Formatters.Binary.__BinaryParser serParser, System.Boolean fCheck) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize (System.IO.Stream serializationStream) [0x00000] in <00000000000000000000000000000000>:0 
  at DavidFDev.DevConsole.DevConsoleData.Load () [0x00000] in <00000000000000000000000000000000>:0 
  at DavidFDev.DevConsole.DevConsoleMono.LoadPreferences () [0x00000] in <00000000000000000000000000000000>:0 
  at DavidFDev.DevConsole.DevConsoleMono.Awake () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Object.Instantiate[T] (T original) [0x00000] in <00000000000000000000000000000000>:0 
  at DavidFDev.DevConsole.DevConsole.Init () [0x00000] in <00000000000000000000000000000000>:0 
UnityEngine.Object:Instantiate(T)
DavidFDev.DevConsole.DevConsole:Init()
```

The error happens only when building with IL2CPP scripting backend, everything seems to be fine when building with Mono scripting backend or if running in the Editor.

**Package version:** 1.0.0
**OS:** Windows 10

**Comment by DavidF-Dev** [comment]
Thanks for raising this issue.

So, to be clear, everything was working fine until you upgraded the Unity version? Also, it works fine when playing in the editor, but throws an exception in builds?

I'm not sure what could have caused this to happen, but I might have an idea for a temporary fix. The package saves its preferences in a binary file in Application.persistentDataPath: https://docs.unity3d.com/ScriptReference/Application-persistentDataPath.html

If you're able to find that file and delete it (devconsole.dat), that may resolve the issue. Let me know how that goes.

When I get time, I will implement and release a fix (will most likely just delete the preferences file if it can't be read, instead of throwing the exception).

On a side note, thanks for using my package! I wasn't sure if anyone was actually using it because it's fairly new. If you have any feedback I would love to hear.

Thanks

**Comment by samlletas** [comment]
That is correct, this started happening only until I upgraded to Unity 2021.2.
Playing in the Editor and builds with Mono scripting backend work fine, but builds with IL2CPP scripting backend throw this exception.

Deleting _devconsole.dat_ file resolves the issue, but after rebuilding the exception is thrown again. I looked at the [Unity release notes](https://unity3d.com/unity/whats-new/2021.2.0?_ga=2.10556912.571432828.1635359211-635649757.1631306898) and it seems to be a know issue in Unity itself:

> IL2CPP: System.Runtime.Serialization.Formatters.BinaryFormatter doesn't work when Project is built with IL2CPP Scripting Backend. ([1374185](https://issuetracker.unity3d.com/issues/il2cpp-system-dot-runtime-dot-serialization-dot-formatters-dot-binaryformatter-doesnt-work-when-project-is-build-with-il2cpp-scripting-backend?_ga=2.48697954.571432828.1635359211-635649757.1631306898))

This is very likely the cause of this exception, and would also explain why the exception is only thrown in IL2CPP builds.

Quite the contrary, thanks to you for creating this package! I have been using it soon after v1.0.0 released and it has been a nice time saver, I love how easy and clean it is to add custom commands. Functionality wise it has been great to me so far.

One thing that could be interesting is to have some way to display custom data in the screen in the same format as how fps are displayed, for example, to show the player current velocity, health, grounded state, etc. Something like the `cs_evaluate` command but in realtime.

**Comment by DavidF-Dev** [comment]
Glad to hear it's Unity's fault and not mine! Nonetheless, I will still catch the exception.

Thanks for your kind words. In regards to your suggestion, I agree that is a very useful feature, however I think that could be achieved on your end via a simple OnGui implementation? I will add it to my list of ideas regardless!

**Comment by samlletas** [comment]
Sounds good!

Oh definitely OnGui works, but I was thinking about having a nicer way to do it with a command to register variables or maybe adding an attribute to the variables themselves and the console taking care of stacking the labels. 

Just an idea that came to my mind to help keep code cleaner, but anyways I think I'm going out of topic 😅. When unity fixes the bug I will post here my findings, thanks for the fast replies!

**Comment by DavidF-Dev** [comment]
Okay, well I've implemented your suggestion via commands, however doing so through an attribute doesn't seem to be possible unless applied to a static field/property. I'll probably release the new version tomorrow - then I will close this issue. Thanks.

**Comment by samlletas** [comment]
All right, sounds good to me then, thanks again.

**Comment by DavidF-Dev** [comment]
Released new version (v1.0.2) with fix for exception upon package initialisation. Use the *stats_help* command for info regarding setting up the tracked variables. Enjoy!

**Comment by samlletas** [comment]
Thanks for the very quick fix! I also played around with the new stats feature and it's seriously awesome! I was not expecting it to be added this soon, this makes it very convenient and easy to display custom data, thank you very much :)
